### PR TITLE
Fixing button behavior, adding software debounce

### PR DESCRIPTION
**Pull request by Ellerbach**
## Description

Fixing button behavior, adding software debounce

## Motivation and Context

- Fixes edge issues with the button and internal timers
- Adding software debounce

## How Has This Been Tested?

on a real big button with an ESP32 :-)

## Types of changes

- [ ] Improvement (non-breaking change that improves a feature, code or algorithm)
- [x] Bug fix (non-breaking change which fixes an issue with code or algorithm)
- [ ] New feature (non-breaking change which adds functionality to code)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Config and build (change in the configuration and build system, has no impact on code or features)
- [ ] Dependencies (update dependencies and changes associated, has no impact on code or features)
- [ ] Unit Tests (add new Unit Test(s) or improved existing one(s), has no impact on code or features)
- [ ] Documentation (changes or updates in the documentation, has no impact on code or features)

## Checklist:

- [x] My code follows the code style of this project (only if there are changes in source code).
- [x] My changes require an update to the documentation (there are changes that require the docs website to be updated).
- [x] I have updated the documentation accordingly (the changes require an update on the docs in this repo).
- [x] I have read the [CONTRIBUTING](https://github.com/nanoframework/.github/blob/main/CONTRIBUTING.md) document.
- [x] I have tested everything locally and all new and existing tests passed (only if there are changes in source code).
- [x] I have added new tests to cover my changes.

### CODE REVIEW

File: devices/Button/ButtonBase.cs [hunk 1]
```diff
@@ -91,17 +93,29 @@ public ButtonBase() : this(TimeSpan.FromTicks(DefaultDoublePressTicks), TimeSpan
         /// </summary>
         /// <param name="doublePress"></param>
         /// <param name="holding"></param>
-        public ButtonBase(TimeSpan doublePress, TimeSpan holding)
+        /// /// <param name="debounceTime">The amount of time during which the transitions are ignored, or zero</param>
```
Comment (mboud) [inline devices/Button/ButtonBase.cs:96]: `/// ///` !

File: devices/Button/ButtonBase.cs [hunk 2]
```diff
@@ -91,17 +93,29 @@ public ButtonBase() : this(TimeSpan.FromTicks(DefaultDoublePressTicks), TimeSpan
...
             _holdingMs = (long)holding.TotalMilliseconds;
+            _debounceTime = debounceTime;
         }
 
         /// <summary>
         /// Handler for pressing the button.
         /// </summary>
         protected void HandleButtonPressed()
         {
+            if (DateTime.UtcNow.Ticks - _debounceStartTicks < _debounceTime.Ticks)
```
[the remark this answers is not in the document]
Comment (mboud) [reply devices/Button/ButtonBase.cs:114]: Nevermind, I mixed ticks and ms, `Environment.TickCount64` is in ms.

File: devices/Button/ButtonBase.cs (hunk 1 above)
Comment (Ellerbach) [reply devices/Button/ButtonBase.cs:96]: good catch...

**Comment by josesimoes** [comment]
Not sure if this is a WIP, but I'm not seeing the Gpio [DebounceTimeout](https://docs.nanoframework.net/api/System.Device.Gpio.GpioPin.html#System_Device_Gpio_GpioPin_DebounceTimeout) property being set... 🤔 

Was under the impression that that was the key part of this fix.

**Comment by Ellerbach** [comment]
Thanks again @mboud for the PR on this PR :-)
@josesimoes if the build is successful, please kindly do a last check and merge if all is ok.

**Comment by mboud** [comment]
> Thanks again @mboud for the PR on this PR :-) @josesimoes if the build is successful, please kindly do a last check and merge if all is ok.

Hi @Ellerbach ! @josesimoes suggested we use the GPIO debounce timeout as well, I totally forgot about that. Should we look into it before merging ?

**Comment by mboud** [comment]
> @mboud glad that you've mentioned it! 😅 And,yes, my question above remains: I was under the impression that the Gpio [DebounceTimeout](https://docs.nanoframework.net/api/System.Device.Gpio.GpioPin.html#System_Device_Gpio_GpioPin_DebounceTimeout) property would be part of the fix. We can revisit that, of course, by my suggestion would be to (at least) expose it in `GpioButton` allowing developers to take advantage of it when using it to interface with mechanical switches.

I'll have a look at it tonight unless Laurent does it in the meantime.

**Comment by Ellerbach** [comment]
> I'll have a look at it tonight unless Laurent does it in the meantime.

Go ahead, no time on my side this week!

**Comment by josesimoes** [comment]
Gentleman: we need to bring a closure to this... 😉

**Review by josesimoes — APPROVED** [review]
LGTM

**Pull request merged**